### PR TITLE
#761 ParallelScheduler does not schedule task evenly between threads

### PR DESCRIPTION
**Pull request by lebannen**
I am not sure about using AtomicIntegerFieldUpdater instead of AtomicInteger.

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/scheduler/ParallelScheduler.java [hunk 1]
```diff
@@ -134,13 +137,7 @@ ScheduledExecutorService pick() {
...
         if (a != SHUTDOWN) {
             // ignoring the race condition here, its already random who gets which executor
-            int idx = roundRobin;
-            if (idx == n) {
-                idx = 0;
-                roundRobin = 0;
-            } else {
-                roundRobin = idx + 1;
-            }
+            int idx = ROUND_ROBIN.getAndUpdate(this, i -> (i < (n-1)) ? i + 1 : 0);
```
Comment (akarnokd) [inline reactor-core/src/main/java/reactor/core/scheduler/ParallelScheduler.java]: From the comment above:

> // ignoring the race condition here, its already random who gets which executor

This is a benign data race we don't care about. The change adds about 20x the overhead for getting a scheduler.

Comment (lebannen) [inline reactor-core/src/main/java/reactor/core/scheduler/ParallelScheduler.java]: Will code like

```
roundRobin++;
if (roundRobin == n) {
  roundRobin = 0;
}
return a[roundRobin];
```

be enough?

Comment (lebannen) [inline reactor-core/src/main/java/reactor/core/scheduler/ParallelScheduler.java]: Or it's possible to make it 
```
int idx = roundRobin;
            if (idx == n) {
                idx = 0;
                roundRobin = 1;
            } else {
                roundRobin = idx + 1;
            }
```

to keep the correct distribution & make minimum changes.

Comment (akarnokd) [inline reactor-core/src/main/java/reactor/core/scheduler/ParallelScheduler.java]: This latter one is okay.

**Comment by Kindrat** [comment]
Actually that is a huge issue. We've faced it on staging when consuming messages from kafka with high load producing. Without thread safety in calculating current index of executor service in array you can schedule multiple tasks for single thread while it is claimed that ParallelScheduler is the most effective for utilising CPU cores.

```
2017-07-27 03:00:55.617 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 238750 ms took processing
2017-07-27 03:00:55.617 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 236010 ms took processing
2017-07-27 03:00:55.618 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 239466 ms took processing
2017-07-27 03:00:55.618 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 239670 ms took processing
2017-07-27 03:00:55.619 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 236827 ms took processing
2017-07-27 03:00:55.620 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235091 ms took processing
2017-07-27 03:00:55.620 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 238347 ms took processing
2017-07-27 03:00:55.621 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 236829 ms took processing
2017-07-27 03:00:55.621 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 236218 ms took processing
2017-07-27 03:00:55.622 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 238045 ms took processing
2017-07-27 03:00:55.622 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 239877 ms took processing
2017-07-27 03:00:55.623 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 237438 ms took processing
2017-07-27 03:00:55.624 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235095 ms took processing
2017-07-27 03:00:55.624 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 240284 ms took processing
2017-07-27 03:00:55.625 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 240689 ms took processing
2017-07-27 03:00:55.626 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 238049 ms took processing
2017-07-27 03:00:55.626 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235816 ms took processing
2017-07-27 03:00:55.627 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 240287 ms took processing
2017-07-27 03:00:55.627 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 237847 ms took processing
2017-07-27 03:00:55.628 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235614 ms took processing
2017-07-27 03:00:55.628 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 237848 ms took processing
2017-07-27 03:00:55.629 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 234692 ms took processing
2017-07-27 03:00:55.630 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 241307 ms took processing
2017-07-27 03:00:55.630 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 240492 ms took processing
2017-07-27 03:00:55.631 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 239277 ms took processing
2017-07-27 03:00:55.632 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 238054 ms took processing
2017-07-27 03:00:55.632 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 236840 ms took processing
2017-07-27 03:00:55.633 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235619 ms took processing
2017-07-27 03:00:55.634 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 234284 ms took processing
2017-07-27 03:00:55.634 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 237449 ms took processing
2017-07-27 03:00:55.635 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 236232 ms took processing
2017-07-27 03:00:55.635 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 234489 ms took processing
2017-07-27 03:00:55.636 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 238058 ms took processing
2017-07-27 03:00:55.637 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235623 ms took processing
2017-07-27 03:00:55.637 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235623 ms took processing
2017-07-27 03:00:55.638 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 236440 ms took processing
2017-07-27 03:00:55.638 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235109 ms took processing
2017-07-27 03:00:55.639 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 234493 ms took processing
2017-07-27 03:00:55.640 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235111 ms took processing
2017-07-27 03:00:55.640 [ INFO] [          parallel-1] c.g.a.b.consumer.KafkaConsumerManager    : 235416 ms took processing
```

and after moving to `Schedulers.fromExecutorService(ForkJoinPool.commonPool())`

```
2017-07-27 05:32:11.153 [ INFO] [.commonPool-worker-3] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-23] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-21] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-12] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-26] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-22] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-18] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-15] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [.commonPool-worker-0] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-30] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-14] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.153 [ INFO] [commonPool-worker-31] c.g.a.b.consumer.KafkaConsumerManager    : 5 ms took processing
2017-07-27 05:32:11.154 [ INFO] [commonPool-worker-24] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [.commonPool-worker-1] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [commonPool-worker-20] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [commonPool-worker-13] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [.commonPool-worker-5] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [commonPool-worker-25] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [commonPool-worker-17] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [.commonPool-worker-4] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [.commonPool-worker-7] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [.commonPool-worker-6] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [commonPool-worker-16] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.154 [ INFO] [commonPool-worker-19] c.g.a.b.consumer.KafkaConsumerManager    : 6 ms took processing
2017-07-27 05:32:11.155 [ INFO] [.commonPool-worker-8] c.g.a.b.consumer.KafkaConsumerManager    : 7 ms took processing
2017-07-27 05:32:11.155 [ INFO] [commonPool-worker-11] c.g.a.b.consumer.KafkaConsumerManager    : 7 ms took processing
2017-07-27 05:32:11.155 [ INFO] [commonPool-worker-27] c.g.a.b.consumer.KafkaConsumerManager    : 7 ms took processing
2017-07-27 05:32:11.155 [ INFO] [commonPool-worker-28] c.g.a.b.consumer.KafkaConsumerManager    : 7 ms took processing
2017-07-27 05:32:11.155 [ INFO] [.commonPool-worker-9] c.g.a.b.consumer.KafkaConsumerManager    : 7 ms took processing
2017-07-27 05:32:11.155 [ INFO] [.commonPool-worker-2] c.g.a.b.consumer.KafkaConsumerManager    : 7 ms took processing
2017-07-27 05:32:11.155 [ INFO] [commonPool-worker-10] c.g.a.b.consumer.KafkaConsumerManager    : 7 ms took processing
```
I'd rather pay even 200x for not catching this in prod.

**Comment by simonbasle** [comment]
@Kindrat would you somehow be able to verify whether the `roundRobin = 1` fix from @lebannen improves that particular situation or not? (eg by building this PR locally)

**Comment by smaldini** [comment]
I think it's reasonable enough to pick worker using stronger consistency @akarnokd if the flows under the hood pick Worker on subscribe. Yes Mono.delay/Flux.interval and other concurrent use of Scheduler.parallel might touch more often but that ensure the advertised fair distribution.

**Comment by smaldini** [comment]
I think @simonbasle we can merge this one and opens a separate one for the stronger consistency using volatile, it might be that index is getting raced but I wonder how is that the case as supposedly this pick doesn't really happen often compared to data flow.

**Comment by smaldini** [comment]
Please @Kindrat try 3.1.0.BS again when ready in 10mn with this fix :)

**Comment by Kindrat** [comment]
@smaldini I'd rather interpret it as a critical bug, man. It's still not working for me b/c index update should be made atomically - the first version of proposed fix. And since it's just a race condition it has even passed our QA and was caught on staging.

Just imagine that on high load processing with multiple cores all of them have received simple int value and incremented it. We are getting random thread locks on 32-core instance all the time with this bug.

**Comment by smaldini** [comment]
@Kindrat how do you reproduce it, or how your flow looks like ? It's interesting to see because the possible race occurs when you really concurrently subscribe many flows (this is when the worker is picked). Is this a case where you flatMap some result and use publishOn inside the flatMap ?

**Comment by smaldini** [comment]
Also I think its worth a new issue indeed.

**Comment by Kindrat** [comment]
in pseudo code:
```
while(true) {
  List<Payload> batch = pollMassiveBatchFromKafka();
  for (Payload payload : batch) {
    Mono.fromRunnable(() -> process(payload))
      .subscribeOn(Schedulers.parallel())
      .subscribe();
  }
}
```

**Comment by akarnokd** [comment]
@Kindrat Unless you run many of such loops in parallel already, you are likely hit by the off-by-one error fixed in this PR.

**Comment by smaldini** [comment]
It might be the case I guess for 
```Mono.fromCallable(() -> random())
             .repeat()
             .flatMap(r -> Mono.fromRunnable(() -> process(payload))
                    .subscribeOn(Schedulers.parallel())
             ).subscribe();```

**Comment by akarnokd** [comment]
You can use `parallel`:

```java
Flux.fromIterable(batch)
.parallel()
.runOn(Schedulers.parallel())
.doOnNext(v -> process(v))
.sequential()
.subscribe();
```

(Maybe a parallel `ignoreElements()` could be added for such side-effect only parallel cases.)

**Comment by smaldini** [comment]
You can already with ParallelFlux#subscribe(), its just running the parallel rails without merge

**Pull request merged**